feat: Add Data Attribute for Cell Tags

jor2 opened the pull request:
## References

Fixes #8298 - Add Data Attribute for Cell Tags

This PR implements the feature requested in the original issue and follows the updated consensus from the discussion in #8410 where it was agreed that all cell tags should result in separate data attributes on the cell (e.g., tag "foo" creates "data-tag-foo" attribute).

## Code changes

- **Added `loadTagsState()` method**: Synchronizes cell tags from metadata with DOM data attributes. Removes existing tag attributes and adds new ones based on current tags.

- **Added `syncTags` property**: A boolean property (defaults to `true`) that controls whether tag synchronization occurs, following the same pattern as existing `syncCollapse` and `syncEditable` properties.

- **Updated `onMetadataChanged()` handler**: Added case for 'tags' metadata changes to trigger `loadTagsState()` when tags are modified.

- **Enhanced `initializeState()` method**: Added conditional call to `loadTagsState()` based on `syncTags` setting during cell initialization.

- **Tag sanitization**: Tag names are sanitized for use as valid HTML data attribute names by replacing non-alphanumeric characters (except hyphens and underscores) with hyphens.

## User-facing changes

- **CSS targeting capability**: Developers can now use CSS selectors to style cells based on their tags (e.g., `[data-tag-parameters]` for papermill parameter cells).

- **JavaScript/DOM access**: The cell tags are now accessible via standard DOM data attribute APIs without needing to access the cell model directly.

- **No visual changes**: This is a purely structural enhancement that enables styling but doesn't change the default appearance of cells.

Example usage:
```css
/* Style cells with "parameters" tag */
[data-tag-parameters] {
  border-left: 4px solid #007acc;
}

/* Style cells with "important" tag */
[data-tag-important] {
  background-color: #fff3cd;
}

Commit:
feat: Add Data Attribute for Cell Tags

jor2 committed:
Merge branch 'main' into fix-tags

jor2 committed:
Merge branch 'main' into fix-tags

jor2 committed:
Merge branch 'main' into fix-tags

jor2 committed:
Merge branch 'main' into fix-tags

krassowski commented on packages/cells/src/widget.ts:
```diff
@@ -492,6 +495,30 @@ export class Cell<T extends ICellModel = ICellModel> extends Widget {
+  /**
+   * Load the tag state from the model.
+   */
+  loadTagsState(): void {
```
Can we make this method private, or is your intent to call it in extensions?


```suggestion
  private _loadTagsState(): void {
```

You could still use `widget['_loadTagsState']()` in tests then.

krassowski commented on packages/cells/src/widget.ts:
```diff
@@ -492,6 +495,30 @@ export class Cell<T extends ICellModel = ICellModel> extends Widget {
+      if (typeof tag === 'string' && tag.length > 0) {
+        // Sanitize tag name for use as data attribute
+        const sanitizedTag = tag.replace(/[^a-zA-Z0-9-_]/g, '-');
+        this.node.setAttribute(`data-tag-${sanitizedTag}`, 'true');
```
Do we need `true` here or can we create data attributes without value instead?

```suggestion
        this.node.setAttribute(`data-tag-${sanitizedTag}`, '');
```

https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute#:~:text=You%20should%20set%20value%20to,necessarily%20do%20what%20you%20expect.

krassowski reviewed (changes requested):
This looks well to me; I left some minor comments and requests on the code.

jor2 committed:
Merge branch 'main' into fix-tags

Commit:
Address PR review comments for tags
  feature

  - Make loadTagsState() private (rename to
  _loadTagsState())
  - Optimize DOM attribute updates to calculate diff
  - Use empty string for data attribute values
  - Change syncTags default to false with explicit enable
   in notebooks

jor2 committed:
Merge branch 'main' into fix-tags
